MetaMask/metamask-extension
# Migrating Icon to typescript and deprecating JS component

## Explanation

Currently the Icon component is built in javascript. This PR deprecates the JS component and creates a new TypeScript version. It also updates all of the import statements that point to the Icon component

- Migrated all `.js` to `.tsx` or `.ts` files
- Followed enum conventions in [TS enum conventions in notion](https://www.notion.so/Use-TypeScript-9ec7bd8fc8cf4d82b81cd8943c4707cd?pvs=4#7e6410b21100455a9a08442108f19d78)
- Created enum version of `ICON_NAMES` => `IconName` as well as keys from UPPERCASE to PascalCase
- Created enum version of `ICON_SIZES` => `IconSize` as well as keys from UPPERCASE to PascalCase
- Updated storybook documentation as well as how to add a new icon

* Part of #18326 
* How to deal with conflicts as a result of this PR https://consensys.slack.com/archives/G1L7H42BT/p1680299226290989

## Screenshots/Screencaps

### Before

https://user-images.githubusercontent.com/8112138/229607315-4838654e-44de-4137-9611-da97ca44f47f.mov

### After

No changes to the functionality of the component, stories and controls all work. Updated docs from objects to enums and how to add a new icon component

https://user-images.githubusercontent.com/8112138/229608558-89e6dd71-f937-4004-b4c6-e7fb8f6a09e9.mov

## Manual Testing Steps

- Go to the storybook build on this PR
- Search `Icon` and go to the `Icon` story in `component-library`
- Check controls work
- Check stories
- Check docs align with enums

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
